components/tidb_query/src/rpn_expr/impl_cast.rs

```diff
@@ -226,55 +377,317 @@ pub fn cast_any_as_decimal<From: Evaluable + ConvertTo<Decimal>>(
+        // TODO, here our `handle_overflow_err` is not same as TiDB's HandleOverflow,
+        //  the latter will return `err` if OverflowAsWarning, but we will return `truncated_wrong_val`.
+        ctx.handle_overflow_err(Error::truncated_wrong_val("INTEGER", origin_str))?;
+        if is_neg {
+            Ok(std::i64::MIN)
+        } else {
+            Ok(std::u64::MAX as i64)
+        }
+    } else {
+        Err(err.into())
+    }
+}
+
+#[rpn_fn(capture = [ctx, extra])]
+#[inline]
+fn cast_decimal_as_uint(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Decimal>,
+) -> Result<Option<i64>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            // TODO, here TiDB round before call `val.is_negative()`
+            if in_union(extra.implicit_args) && val.is_negative() {
+                Ok(Some(0))
+            } else {
+                let r: u64 = val.convert(ctx)?;
+                Ok(Some(r as i64))
+            }
+        }
+    }
+}
+
+#[rpn_fn(capture = [ctx])]
+#[inline]
+fn cast_json_as_uint(ctx: &mut EvalContext, val: &Option<Json>) -> Result<Option<i64>> {
+    match val {
+        None => Ok(None),
+        Some(j) => {
+            let r: u64 = j.convert(ctx)?;
+            Ok(Some(r as i64))
+        }
+    }
+}
+
+// cast any as real, some cast functions reuse `cast_any_as_any`
+//
+// cast_decimal_as_signed_real -> cast_any_as_any<Decimal, Real>
+// cast_time_as_real -> cast_any_as_any<Time, Real>
+// cast_duration_as_real -> cast_any_as_any<Duration, Real>
+// cast_json_as_real -> by cast_any_as_any<Json, Real>
+
+#[rpn_fn]
+#[inline]
+fn cast_signed_int_as_signed_real(val: &Option<Int>) -> Result<Option<Real>> {
+    match val {
+        None => Ok(None),
+        Some(val) => Ok(Real::new(*val as f64).ok()),
+    }
+}
+
 #[rpn_fn(capture = [extra])]
 #[inline]
-fn cast_uint_as_real(extra: &RpnFnCallExtra<'_>, val: &Option<Int>) -> Result<Option<Real>> {
+fn cast_signed_int_as_unsigned_real(
+    extra: &RpnFnCallExtra,
+    val: &Option<Int>,
+) -> Result<Option<Real>> {
     match val {
         None => Ok(None),
         Some(val) => {
-            // TODO, TiDB's here may has bug(val is uint, why it will <0 ?),
-            // fix this after TiDB's had fixed.
             if in_union(extra.implicit_args) && *val < 0 {
-                return Ok(Some(Real::new(0f64).unwrap()));
+                Ok(Some(Real::zero()))
+            } else {
+                // FIXME, TiDB here may has bug, fix this after fix TiDB's
+                Ok(Real::new(*val as u64 as f64).ok())
+            }
+        }
+    }
+}
+
+#[rpn_fn]
+#[inline]
+fn cast_unsigned_int_as_signed_or_unsigned_real(val: &Option<Int>) -> Result<Option<Real>> {
+    match val {
+        None => Ok(None),
+        Some(val) => Ok(Real::new(*val as u64 as f64).ok()),
+    }
+}
+
+#[rpn_fn]
+#[inline]
+fn cast_real_as_signed_real(val: &Option<Real>) -> Result<Option<Real>> {
```

Comment:
Can we use a generic function for `cast_int_as_int` (https://github.com/tikv/tikv/pull/5357/files#diff-7b0a1830a6f61d6aedda123d5086c8edR391) and `cast_real_as_signed_real `?

Reply:
I think it is unnecessary to add `convertto<T> for T`

Reply:
Not add `ConvertTo<T> for T`, but a new RPN function which used to cast to the same type.

Reply:
`string_as_string`, `real_as_real`, `decimal_as_decimal`, `time_as_time` has special logic, so i think it is unnecessary to add this generic function.